### ISSUE TITLE
question about pipeline and zero stage[question]

### ISSUE THREAD

**Issue by etoilestar** [microsoft/DeepSpeed#4164]
hello, this is not a bug report. I have been studying your code recently, but I don't understand why the pipeline model cannot be used when using zero_stage>1. This is judged as an assertion statement in the code https://github.com/microsoft/DeepSpeed/blob/bd65eeaf708dfab70e372642284c6af06d506258/deepspeed/runtime/pipe/engine.py#L70. When I comment on this statement, the error is reported as The gradient is calculated twice. When I read the code, I didn't see how the gradient is split, please give me some advice on how to learn your code, thank you.

**Comment by ShadenSmith** [comment]
Hi @etoilestar! There's no fundamental reason why pipeline parallelism and the later stages of zero couldn't be combined, but they take opposing paths to optimizations and are likely to hurt each other's performance. That's why we only integrated zero stage 1. I'm not sure of the source of the crash after removing the assertion, but I imagine there could be some work required to integrate the systems more fully.

Pipeline parallelism relies on lots of micro-batches for efficiency -- preferably at least 8x the degree of pipeline parallelism. The later stages of zero introduce data-parallel communications during the forward and backward passes. These communications would need to be performed 10s-100s of times per training step, and the weights/gradients being communicated are much large than the activations communicated by pipeline parallelism each micro-batch.
